File: 백준/Gold/21608. 상어 초등학교/상어 초등학교.py
import sys
input = lambda: sys.stdin.readline().rstrip()

N = int(input())
D = {i[0]: i[1:] for i in [list(map(int, input().split())) for _ in range(N * N)]}
A = [[0 for _ in range(N)] for _ in range(N)]
dr, dc = (-1, 1, 0, 0), (0, 0, -1, 1)
for k, v in D.items():
    pos1 = []
    maxCnt = 0
    for r in range(N):
        for c in range(N):
            if A[r][c] == 0:
                cnt = 0
                for j in range(4):
                    mr, mc = r + dr[j], c + dc[j]
                    if 0 <= mr <= N - 1 and 0 <= mc <= N - 1 and A[mr][mc] in v:
                        cnt += 1
                if maxCnt < cnt:
                    maxCnt = cnt
                    pos1 = [[r, c]]
                elif maxCnt == cnt:
                    pos1.append([r, c])
    if len(pos1) == 1:
        A[pos1[0][0]][pos1[0][1]] = k
        continue
    pos2 = []
    maxCnt = 0
    for r, c in pos1:
        cnt = 0
        for j in range(4):
            mr, mc = r + dr[j], c + dc[j]
            if 0 <= mr <= N - 1 and 0 <= mc <= N - 1 and A[mr][mc] == 0:
                cnt += 1
        if maxCnt < cnt:
            maxCnt = cnt
            pos2 = [[r, c]]
        elif maxCnt == cnt:
            pos2.append([r, c])
    if len(pos2) == 1:
        A[pos2[0][0]][pos2[0][1]] = k
        continue
    pos3 = [float('inf'), float('inf')]
    for r, c in pos2:
        if r < pos3[0]:
            pos3 = [r, c]
        elif r == pos3[0]:
            pos3[1] = min(pos3[1], c)
    A[pos3[0]][pos3[1]] = k
S = [0, 1, 10, 100, 1000]
res = 0
for r in range(N):
    for c in range(N):
        cnt = 0
        for i in range(4):
            mr, mc = r + dr[i], c + dc[i]
            if 0 <= mr <= N - 1 and 0 <= mc <= N - 1 and A[mr][mc] in D[A[r][c]]:
                cnt += 1
        res += S[cnt]
print(res)

File: 백준/Gold/21608. 상어 초등학교/상어 초등학교.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		int N = Integer.parseInt(br.readLine());
		HashMap<Integer, int[]> map = new LinkedHashMap<>();
		for (int i = 0; i < N * N; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			int key = Integer.parseInt(st.nextToken());
			int[] arr = new int[4];
			for (int j = 0; j < 4; j++) {
				arr[j] = Integer.parseInt(st.nextToken());
			}
			map.put(key, arr);
		}
		int[][] A = new int[N][N];
		
		int[] dr = {-1, 1, 0, 0};
		int[] dc = {0, 0, -1, 1};
		for (int k : map.keySet()) {
			ArrayList<int[]> pos1 = new ArrayList<>();
			int maxCnt = 0;
			for (int r = 0; r < N; r++) {
				for (int c = 0; c < N; c++) {
					if (A[r][c] == 0) {
						int cnt = 0;
						for (int i = 0; i < 4; i++) {
							int mr = r + dr[i], mc = c + dc[i];
							if (mr >= 0 && mr < N && mc >= 0 && mc < N) {
								int[] arr = map.get(k);
								for (int j = 0; j < 4; j++) {
									if (A[mr][mc] == arr[j]) {
										cnt++;
										break;
									}
								}
							}
						}
						if (maxCnt < cnt) {
							maxCnt = cnt;
							pos1.clear();
							pos1.add(new int[] {r, c});
						}
						else if (maxCnt == cnt) {
							pos1.add(new int[] {r, c});
						}
					}
				}
			}
			if (pos1.size() == 1) {
				int[] rc = pos1.get(0);
				A[rc[0]][rc[1]] = k;
				continue;
			}
			
			ArrayList<int[]> pos2 = new ArrayList<>();
			maxCnt = 0;
			for (int[] rc: pos1) {
				int r = rc[0], c = rc[1];
				int cnt = 0;
				for (int i = 0; i < 4; i++) {
					int mr = r + dr[i], mc = c + dc[i];
					if (mr >= 0 && mr < N && mc >= 0 && mc < N && A[mr][mc] == 0) {
						cnt++;
					}
				}
				if (maxCnt < cnt) {
					maxCnt = cnt;
					pos2.clear();
					pos2.add(new int[] {r, c});
				}
				else if (maxCnt == cnt) {
					pos2.add(new int[] {r, c});
				}
			}
			if (pos2.size() == 1) {
				int[] rc = pos2.get(0);
				A[rc[0]][rc[1]] = k;
				continue;
			}
			
			int[] pos3 = {Integer.MAX_VALUE, Integer.MAX_VALUE};
			for (int[] rc: pos2) {
				int r = rc[0], c = rc[1];
				if (r < pos3[0]) {
					pos3[0] = r;
					pos3[1] = c;
				}
				else if (r == pos3[0]) {
					pos3[1] = Math.min(pos3[1], c);
				}
			}
			A[pos3[0]][pos3[1]] = k;
		}
		
		int[] s = {0, 1, 10, 100, 1000};
		int result = 0;
		for (int r = 0; r < N; r++) {
			for (int c = 0; c < N; c++) {
				int cnt = 0;
				for (int i = 0; i < 4; i++) {
					int mr = r + dr[i], mc = c + dc[i];
					if (mr >= 0 && mr < N && mc >= 0 && mc < N) {
						int[] arr = map.get(A[r][c]);
						for (int j = 0; j < 4; j++) {
							if (A[mr][mc] == arr[j]) {
								cnt++;
								break;
							}
						}
					}
				}
				result += s[cnt];
			}
		}
		System.out.println(result);
	}
}